data_ingest/ingestors.py

```
@@ -244,18 +244,53 @@ def formatted(self, source, unformatted):
-        'error_columns': [
+    error = {}
+    error['severity'] = rule.get('severity', 'Error')
+    error['code'] = rule.get('error_code')
+    error['error_columns'] = [
             k for (idx, k) in enumerate(row_dict.keys())
             if k in rule['columns']
         ]
-    }
+
+    # create message
+    message = rule.get('message', '')
+    # Pattern that will match everything that looks like this: {...}
+    pattern = re.compile(r'\{.*?\}')
+    fields = pattern.findall(message)
+    for field in fields:
+        # Remove { }
+        key = field[1:-1].strip()
+        # Direct Substitution
+        if key in row_dict.keys():
+            message = message.replace(field, row_dict[key])
+        # Expression Calculation and Substitution
+        else:
+            # This will put out the two field names (strip out any spaces), and the operator
+            # (operand1 operator operand2)
+            expression = re.match(r'^\s*(.*?)\s*([\+\-\*/])\s*(.*?)\s*$', key)
+            try:
+                operand1, operator, operand2 = expression.groups()
+
+                value1 = RowwiseValidator.cast_value(row_dict.get(operand1, ''))
+                value2 = RowwiseValidator.cast_value(row_dict.get(operand2, ''))
+
+                # only supporting int/float operations
+                supported_type = (float, int)
+                if any(isinstance(value1, t) for t in supported_type) and \
+                   any(isinstance(value2, t) for t in supported_type):
+                    # Only using a controlled set of operations here, the operation is
+                    # well-examined to know it doesn't create any security issue
+                    result = eval(f'{value1} {operator} {value2}')
+                    message = message.replace(field, str(result))
+
+            except AttributeError:
+                # This means the expression is malformed, will not replace field with anything
+                continue
```

Should we raise some sort of validation error here rather than failing silently?

---

That's right, right now it just doesn't replace anything, but yes it is a good idea to notify the rules maintainer so they know maybe they misspelled something or whatever it is.  I will rethink where that should happen, and that probably means changing how the current error structures look like.